_includes/v19.1/app/insecure/gorm-sample.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"fmt"
+	"log"
+	"math"
+	"math/rand"
+	"time"
+
+	// Import GORM-related packages.
+	"github.com/jinzhu/gorm"
+	_ "github.com/jinzhu/gorm/dialects/postgres"
+)
+
+// Account is our model, which corresponds to the "accounts" database table.
+type Account struct {
+	ID      int `gorm:"primary_key"`
+	Balance int
+}
+
+func main() {
+	// Connect to the "bank" database as the "maxroach" user.
+	const addr = "postgresql://maxroach@localhost:26257/bank?sslmode=disable"
+	db, err := gorm.Open("postgres", addr)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer db.Close()
+
+	db.LogMode(false)
+
+	// Automatically create the "accounts" table based on the Account
+	// model.
+	db.AutoMigrate(&Account{})
+
+	// Insert two rows into the "accounts" table.
+	db.Create(&Account{ID: 1, Balance: 1000})
+	db.Create(&Account{ID: 2, Balance: 250})
+
+	printBalances(db)
+
+	// Transfer funds between accounts. To handle any possible
+	// transaction retry errors, we add a retry loop with exponential
+	// backoff to the transfer logic (see below).
+
+	var maxRetries = 3
+	var amount = 100
+
+	var fromAccount Account
+	var toAccount Account
+
+	db.First(&fromAccount, 1)
+	db.First(&toAccount, 2)
+
+retryLoop:
+	for retries := 0; retries <= maxRetries; retries++ {
+
+		if retries == maxRetries {
+			fmt.Errorf("hit max of %d retries, aborting", retries)
+			break retryLoop
+		}
+
+		err := transferFunds(db, fromAccount, toAccount, amount)
+
+		if err != nil {
+			// If there's an error, we try again after sleeping an
+			// increased amount of time, a.k.a. exponential backoff.
+			fmt.Println("error: ", err)
+			r := rand.New(rand.NewSource(time.Now().UnixNano()))
+			var sleepMs = ((int(math.Pow(2, float64(retries))) * 1000) + r.Intn(100-1) + 1)
+			fmt.Printf("sleepMs = %s\n", sleepMs)
+			time.Sleep(time.Millisecond * time.Duration(sleepMs))
+		} else {
+			break retryLoop
+		}
+	}
+
+	// Print balances after transfer to ensure that it worked.
+	printBalances(db)
+
+	// Delete accounts so we can start fresh when we want to run this
+	// program again.
+	deleteAccounts(db)
+}
+
+func transferFunds(db *gorm.DB, fromAccount Account, toAccount Account, amount int) error {
+	if fromAccount.Balance < amount {
+		err := fmt.Errorf("account %d balance %d is lower than transfer amount %d", fromAccount.ID, fromAccount.Balance, amount)
+		return err
+	}
+
+	// We would like to check for the 40001 error code below, but it
+	// isn't clear how to get at it from GORM. According to
+	// https://github.com/jinzhu/gorm/issues/17, GORM does not expose
+	// the PG error codes. Therefore, we retry on any error.
+
+	db.Exec("BEGIN")
```

Comment:
Transaction statements like BEGIN are special and can't be executed on a plain `DB` object. You have to use `txn := db.Begin()` and do the rest of the statements on `txn` instead of `db` (and similarly for the `Commit` and `Rollback` methods).

Reply:
> Transaction statements like BEGIN are special and can't be executed on a plain DB object. You have to use txn := db.Begin() and do the rest of the statements on txn instead of db (and similarly for the Commit and Rollback methods).

I tried changing everything to use `txn := db.Begin()` and `txn.Foo()` everywhere, but the code no longer works in that I could no longer see the explicit `BEGIN` etc. statements being passed to the DB, nor could I cast the Error object and get the return code.  I pasted the code and output to https://pastebin.com/ELtLu5yk

Reply:
`db.Exec("BEGIN")` will work as long as you only have one thread, but it's not safe with multiple threads because of the way connection pools work (unless gorm's connection pool is significantly different from `database/sql`'s). So we do need to get to the bottom of this.

It looks like you are successfully casting the error to `pq.Error`; if that was failing you'd get a panic.

Reply:
> db.Exec("BEGIN") will work as long as you only have one thread, but it's not safe with multiple threads because of the way connection pools work (unless gorm's connection pool is significantly different from database/sql's). So we do need to get to the bottom of this.

OK.  I'll revise to use `txn = db.Begin()` and friends and see if I can find the incantations it wants to make it work with retry errors.